be/src/io/shared_buffered_input_stream.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+
+#include "common/status.h"
+#include "io/seekable_input_stream.h"
+
+namespace starrocks::io {
+
+class SharedBufferedInputStream : public SeekableInputStream {
+public:
+    struct IORange {
+        int64_t offset;
+        int64_t size;
+        bool operator<(const IORange& x) const { return offset < x.offset; }
+    };
+    struct CoalesceOptions {
+        static constexpr int64_t MB = 1024 * 1024;
+        int64_t max_dist_size = 1 * MB;
+        int64_t max_buffer_size = 8 * MB;
+    };
+
+    SharedBufferedInputStream(std::shared_ptr<SeekableInputStream> stream, const std::string& filename, size_t size);
+    ~SharedBufferedInputStream() override = default;
+
+    Status seek(int64_t position) override {
+        _offset = position;
+        return Status::OK();
+    }
+    StatusOr<int64_t> position() override { return _offset; }
+    StatusOr<int64_t> read(void* data, int64_t count) override;
+    Status read_at_fully(int64_t offset, void* out, int64_t count) override;
+    StatusOr<int64_t> get_size() override;
+    Status skip(int64_t count) override {
+        _offset += count;
+        return Status::OK();
+    }
+
+    Status set_io_ranges(const std::vector<IORange>& ranges);
+    void release_to_offset(int64_t offset);
+    Status get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes);
+    void release();
+    void set_coalesce_options(const CoalesceOptions& options) { _options = options; }
```

Comment:
why not pass this in ctor?

Reply:
because this parameter is optional, there is default setting in this class.